### PR TITLE
NARA Mapper first pass

### PR DESCRIPTION
**Pull request by mdellabitta**
There's a few TODOs and things I'll need to talk to Gretchen about. 
Ran this locally on my laptop and it was done in well under an hour.
Codacy will probably yell at me, and I'll work on what it says.

### CODE REVIEW

**Comment by mdellabitta** [comment]
Clearly needs a rebase on the date fix.

**Comment by mdellabitta** [comment]
Did a rebase to get the `develop` in there, but didn't squash yet.

File: src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    f"""$value3"; $value2"; "$value1""""
+  }
+
+  def rights(xml: NodeSeq): Seq[String] = for {
+    useRestriction <- xml \ "useRestriction"
+    value1 = (useRestriction \ "note").text
+    value2 = (useRestriction \ "specificUseRestrictionArray" \ "specificUseRestriction" \ "termName").text
+    value3 = (useRestriction \ "status" \ "termName").text
+  //VALUE2": "VALUE1" "VALUE3
+  } yield f"""$value2": "$value1" "$value3"""
```
Comment (markbreedlove) [inline src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala]: Are these double-quotes correct? They appear not to match the pattern in the comment above.

File: src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+      coverageDate <- xml \\ "coverageDates"
+      coverageStartDate = (coverageDate \ "coverageStartDate").headOption
+      coverageEndDate = (coverageDate \ "coverageEndDate").headOption
+      //dateQualifier = coverageDate \ "dateQualifier" //todo not sure what to do with qualifier in EDTF
+      if coverageStartDate.nonEmpty || coverageEndDate.nonEmpty
+
+    } yield {
+      val displayDate = getDisplayDate(coverageStartDate, coverageEndDate)
+      EdmTimeSpan(
+        originalSourceDate = displayDate, //this is cheating?
```
Comment (markbreedlove) [inline src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala]: `sourceResource.date` is supposed to be for date of publication or date of creation, as I understand it, whereas `sourceResourc.temporal` is for "aboutness." Therefore, with regard to whether this is cheating, I'd just say it's not if you can confirm that these "coverage dates" are for the date the source resource was generated or published.

File: src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+import dpla.ingestion3.mappers.rdf.DCMIType
+import dpla.ingestion3.mappers.xml.XmlExtractionUtils
+import dpla.ingestion3.model._
+import org.eclipse.rdf4j.model.IRI
+
+import scala.xml.{Node, NodeSeq, XML}
+
+class NaraExtractor(rawData:String) extends Extractor with XmlExtractionUtils with Serializable {
+
+  implicit val xml: NodeSeq = XML.loadString(rawData)
```
Comment (markbreedlove) [inline src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala]: Could you comment on why this `implicit val xml` is an implicit val?

I see that `dpla.ingestion3.mappers.xml.XmlExtractionUtils` has the methods `extractString` and `extractStrings`, which each have an `implicit val xml`, but I have the sense I'm missing something when I read this.

Comment (mdellabitta) [inline src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala]: Is there something I should comment or explain?

Comment (markbreedlove) [inline src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala]: I guess just why not make it `val xml`?

File: src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala (hunk 1 above)
Comment (mdellabitta) [inline src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala]: They seem to match to me, but maybe I'm taking the notation too literally?

Note that the triple quotes at the beginning and end mean that single quotes no longer require escaping. Also note that I have no idea what's up with the specification in the comment, and I could be implementing it entirely incorrectly.

File: src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala (hunk 3 above)
Comment (mdellabitta) [inline src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala]: It means I don't need to pass it explicitly to every method, which reduces noise. We've followed this pattern in the other mappers.

File: src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala (hunk 1 above)
Comment (markbreedlove) [inline src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala]: Oh, right, I guess I didn't look at that closely.
Still, as you mentioned, I don't get the unclosed double-quotes in the specification.

File: src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+      org <- xml \\ "organizationalContributorArray" \ "organizationalContributor"
+      name = (org \ "contributor" \ "termName").text
+      _type = (org \ "contributorType" \ "termName").text
+      if _type != "Publisher"
+    } yield name
+
+    val personalContributors = for {
+      person <- xml \\ "personalContributorArray" \ "personalContributor"
+      name = (person \ "contributor" \ "termName").text
+      //_type = (person \ "contributorType" \ "TermName").text
```
Comment (AudreyAltman) [inline src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala]: Do you want to keep the commented-out line above?

File: src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+    val releaseDates = simpleDate(xml \\ "releaseDateArray" \ "proposableQualifiableDate")
+
+    Seq(
+      coverageDates,
+      copyrightDates,
+      productionDates,
+      broadcastDates,
+      releaseDates
+    )
+      .collectFirst({ case x => x })
```
Comment (AudreyAltman) [inline src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala]: I'm not clear what this `collectFirst` method is doing.

Comment (mdellabitta) [inline src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala]: It gets the first result that's not None. So, the first execution that returns a value. There's a better way to do this, though, so I'll check that in instead.

File: src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala (hunk 4 above)
Comment (mdellabitta) [inline src/main/scala/dpla/ingestion3/mappers/providers/NaraExtractor.scala]: It's mentioned in the spec, but doesn't do anything. Not quite sure what to do with it, to be honest!

**Pull request merged**